python/lsst/analysis/drp/makeQaTractTables.py

```diff
@@ -0,0 +1,326 @@
+                      format(filterName, tract, len(inputObjCats)))
+        # It is much faster to concatenate a list of DataFrames than to
+        # concatenate successively within the for loop.
+        catList = []
+        colsToLoadList = None
+        refColsToLoadList = None
+        measColsToLoadList = None
+        for inputCat in inputObjCats:
+            patch = inputCat.dataId["patch"]
+            inputCatCols = inputCat.get(component="columns")
+            inputCatFilters = inputCatCols.get_level_values("filter").unique()
+            inputCat = inputCat.get()  # Required because deferLoad=True was set
+            haveFilter = True
+            tempFilterName = filterName
+            if filterName not in inputCatFilters:
+                patchId = inputCat["meas"][inputCatFilters[0]]["patchId"].unique()[0]
+                self.log.info("Filter {} does not exist for: {}, {:2d} [{}].  Setting NaNs for patch "
+                              "columns...".format(filterName, tract, int(patch), patchId))
+                haveFilter = False
+                tempFilterName = inputCatFilters[0]
+            measCat = inputCat["meas"][tempFilterName]
+            if self.config.dfDataset == "forced":
+                forcedCat = inputCat["forced_src"][tempFilterName]
+                if colsToLoadList is None:
+                    colsToLoadList = [col for col in forcedCat.columns if
+                                      (col.startswith(tuple(self.config.baseColStrList))
+                                       and not any(s in col for s in self.config.notInColStrList))]
+                cat = forcedCat[colsToLoadList]
+                # Insert select columns from the ref and meas cats for the
+                # forced cats.
+                refCat = inputCat["ref"][tempFilterName]
+                if refColsToLoadList is None:
+                    refColsToLoadList = [col for col in refCat.columns if
+                                         (col.startswith(tuple(self.config.columnsToCopyFromRef))
+                                          and not any(s in col for s in self.config.notInColStrList))]
+                refCat = refCat[refColsToLoadList]
+                if measColsToLoadList is None:
+                    measColsToLoadList = [col for col in measCat.columns if
+                                          (col.startswith(tuple(self.config.columnsToCopyFromMeas))
+                                           and not any(s in col for s in self.config.notInColStrList))]
+                measCat = measCat[measColsToLoadList]
+                if not haveFilter:  # Fill in missing patch columns win NaNs
+                    cat = pd.DataFrame().reindex_like(cat)
+                    refCat = pd.DataFrame().reindex_like(refCat)
+                    measCat = pd.DataFrame().reindex_like(measCat)
+                cat = pd.concat([cat, refCat, measCat], axis=1)
```

Comment:
How about we move the code in this `if` block into a method that returns `cat` and takes whatever of the local variables it uses as arguments?  that can be an abstract method that raises `NotImplementedError` in the base class, have the above code in `MakeForced...` and code below in `MakeUnforced...`.

There might be a lot of arguments at first, and maybe more than just `cat` returned, so this may not be the final form, but I bet it'll be closer, and it'll make it clear what step to take next.